Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+```hs
+str :: Parser String
+str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))
+-- Or:
+--- str = label "string" $ char '"' *> takeWhileP Nothing (/= '"') <* char '"'
+```
+
+In `atom`, include `SString <$> str`.
+
+```hs
+>>> parseTest atom "\"hey\""
+SString "hey"
+```
+
+### Identifiers
+
+An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the one that is frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.
+
+Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `many` tries to run the given parser **0 or more times**. In regex, our `identifier` parser is equivalent to `[a-zA-Z_][a-zA-Z0-9_]*`.
+
+```hs
+identifier :: Parser Identifier
+identifier = label "identifier" $ do
+  first <- letterChar <|> char '_'
+  rest <- many $ alphaNumChar <|> char '_'
+  pure $ Identifier $ first : rest
+-- Or:
+-- identifier = Identifier <$> label "identifier" ((:) <$> (letterChar <|> char '_') <*> many (alphaNumChar <|> char '_'))
+```
+
+Now include `SId <$> identifier` to `atom`.
+
+**Warning**: When parsing with `<|>` or `choice`, **the order of the parsers matters**. Take this as an example:
+
+```hs
+>>> parseTest (SBool <$> bool <|> SId <$> identifier) "true"
+SBool True
+
+>>> parseTest (SId <$> identifier <|> SBool <$> bool) "true"
+SId (Identifier {getId = "true"})
+```
+
+In the first case, we parse a boolean before an identifier, which causes `true` to be matched by `bool`. However, in the second example, using `identifier` before `bool` caused `true` to be recognized as an identifier. You should be careful while considering such things. In most cases, you can assume that parsers are greedy. Notice that in the second case, the `bool` parser will never be executed.
+
+In other words, make sure that you parse a boolean in `atom` before you parse an identifier.
+
+### S-expressions
+
+Finally, let's move on to our last parser, which parses S-expressions themselves, such as `(foo 42 "hey")` or `(bar)`.
+
+First, we will do it the naïve way, which may come as the first natural solution to someone just learning Megaparsec, and then we will do it in a better manner using extra tools offered by Megaparsec.
+
+#### S-expressions: The awkward way
+
+The idea here is to parse an atom, followed by 0 or more atoms. Of course, this means that we want to use `many` to parse such atoms. Additionally, an S-expression may be surrounded by parenthesis, so we can use `between` for such a task.
+
+This parser, however, is not as simple as before, but let's first do it the wrong way (which won't work properly), analyze the problem, and come up with a solution to it.
+
+```hs
+sexp :: Parser (SExp, [SExp])
+sexp = label "S-expression" $ between (char '(') (char ')') ((,) <$> atom <*> many atom)
```

Comment:
Isn't that `Control.Applicative.some`?

Ah, I get it: nobody made `some` return `Data.List.NotEmpty.NotEmpty` yet.

Reply:
Actually, there is: https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Applicative-Combinators-NonEmpty.html#v:some

But Megaparsec doesn't export this function and instead exports the possibly-empty-list one. But I added a small note about it.